addons/mail/models/mail_activity.py

```diff
@@ -504,11 +504,12 @@ def action_feedback_schedule_next(self, feedback=False):
             'target': 'new',
         }
 
-    def _action_done(self, feedback=False, attachment_ids=None):
+    def _action_done(self, feedback=False, attachment_ids=None, new_obj=None):
```

Comment:
Hello,

That's a bit hacking the API for a single use case. That kind of constructs leads to a ping-pong style code with a hard-to-understand API.

I would instead delegate the next activities values computation to a sub-method that you could override in a clean way in documents. Seems this method is not intended to be used as it is currently (especially documents override which is quite... weird). We should begin cleaning it and using sub methods is probably better.

Could you give it a try ?

By the way, unit tests will be required in test_mail module as for each modification in mail code.

Feel free to ask id anything is unclear !

Many thanks,

Reply:
Indeed, I agree about the code being ... "weird"
I'll take a stab at improving it and I'll drop you a note if I need assistance

Reply:
Hi @tde-banana-odoo 

I have extracted the next activities values preparation into a separate method. I have it in 2 commits as I no longer require this. The document creation is now done in the `create()` override and not in `_action_done()` - so it is a little cleaner than the previous hack. refer odoo/enterprise#20656.

Do you still think it will be better to create the new document in a `_prepare_next_activity_values` override? I don't see much of a difference between the 2

I have a question related to the enterprise commit... the create method uses a list of new document values to create them (even though it is unlikely to ever create multiple documents) - I require the ID of the new document to link to the mail.activity - what would be a better way to do this? This is the only reason I can think of to override `_prepare_next_activity_values`